New Lookup Table uploader

millerdev opened the pull request:
This PR implements a new Lookup Table (aka Fixture) uploader that replaces the existing two uploader implementations. The new uploader does not change tables, rows, and/or ownership records that did not change. Ownership assignments are MUCH more efficient on account of many db hits being replaced by bulk user, group, and location queries. The new uploader only performs insert and delete operations on the database, which will make it easier to maintain performance during the pending Couch to SQL Lookup Tables migration.

:tropical_fish: Review by commit.

## Safety Assurance

### Safety story

All features were implemented with TDD, and tests were structured such that they were run against the old and new uploaders throughout the development process. Only in the final commit were the old uploaders removed along with the tests that were used to verify feature parity.

### Automated test coverage

Yes.

### QA Plan

Planning to deploy to staging and request a QA pass on the lookup table uploader.
https://dimagi-dev.atlassian.net/browse/QA-4281

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

millerdev committed:
Refactor and clean up lookup table uploader tests

millerdev committed:
Run lookup table uploader tests on fast uploader

Includes a cache invalidation bugfix for the fast uploader.

millerdev committed:
Implement row replacement in new uploader

millerdev committed:
Replace duplicate rows in new uploader

millerdev committed:
Preserve row order in new uploader

millerdev committed:
Implement table deletion in new uploader

millerdev committed:
Move item update logic into get_mutation()

millerdev committed:
New lookup table uploader

This is intended to eventually replace both of the existing uploader implementations with something that will be fast and easier to migrate to work with SQL Lookup Tables. It works with existing uploader test, but is nowhere near complete.

millerdev committed:
Implement table update in new uploader

millerdev committed:
Implement row deletion in new uploader

millerdev committed:
Always use existing row sort keys

Only has an effect when replacing a table, and can result in less row churn.

millerdev committed:
Implement basic row ownership in new uploader

Error handling for missing owners is not implemented.

millerdev committed:
Improve test setup and teardown

millerdev committed:
Refactor test upload helper to accept workbook

millerdev committed:
Implement upload progress and some error checks

The errors checked are reported by the old uploader, but not the new one.

millerdev committed:
Improve ownership name matching and error handling

millerdev committed:
Handle invalid username in new uploader

millerdev committed:
Handle non-string owner names (old and new uploader)

millerdev committed:
Implement fast upload features in new uploader

millerdev committed:
Add fields to table_key and row_key; test keys

millerdev committed:
Chunk bulk saves in CouchTransaction

millerdev committed:
Replace old uploaders with new

millerdev committed:
make translations

Note: running `make translations` locally did not produce the correct output (line wrapping was wrong), possibly because of an incompatibility in my local gettext libraries? The contents of this commit were applied from the diff output in the failed Github Actions test run.

calellowitz reviewed (commented):
Generally looks good, few questions. Apologies if some of them are answered in later commits, I tried to delete all the comments that no longer applied but may have missed a couple.

In addition to the QA pass you mentioned in the description, it might be nice to do some performance testing vs both the regular and fast uploader so that we have sense of the expected improvement.

millerdev commented:
> it might be nice to do some performance testing vs both the regular and fast uploader so that we have sense of the expected improvement.

Yep, performance testing is included in the [QA ticket](https://dimagi-dev.atlassian.net/browse/QA-4281?focusedCommentId=212454).

millerdev commented on corehq/apps/fixtures/tests/test_upload.py:
```diff
@@ -329,7 +333,7 @@ def get_workbook_from_data(self, headers, rows):
         data = self.make_rows(rows)
         workbook = self.get_workbook_from_data(self.headers, data)
-        _run_fixture_upload(self.domain, workbook)
+        type(self).do_upload(self.domain, workbook)
```
`do_upload` is a function assigned to the class. Accessing it as `self.do_upload` causes it to be "bound" to `self`, which causes `self` to be passed as the first (`domain`) argument.

calellowitz replied:
Hm, that doesn't match the [python documentation](https://docs.python.org/3/library/functions.html#staticmethod) (or my experience). Static methods should be able to be called fine on instances of a class rather than just the class itself.

I did a test of that myself
```
>>> class A:
...     @staticmethod
...     def b(var):
...         print(var)
... 
>>> A.b('hi')
hi
>>> A().b('hi')
hi
```

millerdev commented on corehq/apps/fixtures/upload/run_upload.py:
```diff
@@ -52,6 +52,8 @@ def _run_upload(domain, workbook, replace=False, task=None):
     def process_table(table, new_table):
         def process_row(row, new_row):
             update_progress(table)
+            if skip_orm:
```
The `skip_orm` parameter references the feature flag that enables the "fast" fixture uploader.

https://github.com/dimagi/commcare-hq/blob/033796fc2867f0c4f4f23d1a67bbcd429b00b471/corehq/toggles/__init__.py#L1806-L1808

That language is now less accurate than it was before, and once Lookup Tables are migrated to SQL that language will be even less meaningful, but changing it seems out of scope for the migration. Rather than change it, I would prefer to remove that feature flag since I think the benefits will be negligible with the new uploader. I will wait for the results of performance testing to decide how to proceed.

calellowitz replied:
I am so curious what that person meant, but agree it is definitely not worth changing.

millerdev commented on corehq/apps/fixtures/upload/workbook.py:
```diff
@@ -36,21 +38,31 @@ def __init__(self, file_or_filename):
-    def get_data_sheet(self, data_type_tag):
-        return self.workbook.get_worksheet(data_type_tag)
+    def get_data_sheet(self, tag):
+        # Cache all rows in memory in order to traverse them more than
```
I doubt it since rows are usually relatively small objects. The largest lookup tables spreadsheet I have seen is 20MB (from one of the domains for which `SKIP_ORM_FIXTURE_UPLOAD` is enabled). I plan to do some memory profiling with that to verify that this is not a problem.

millerdev replied:
Additionally, without this I could not think of another way to efficiently load owner ids in bulk (rather than on demand as each rows is processed). Currently the old uploader does something like 6 db hits per row while processing ownership. Granted, some of that could be optimized, but it is difficult to make it efficient without multiple passes per sheet.

calellowitz replied:
Sounds good.

millerdev replied:
I've done some memory profiling against that 20MB spreadsheet, and while I have not precisely measured the memory impact of caching rows like this, I can confirm that this is by far not the most significant consumer of memory in the uploader. Accumulating `FixtureDataType`, `FixtureDataItem` and `FixtureOwnership` objects in the `CouchTransaction` and then saving them to Couch in bulk takes MUCH more memory (like over 10x).

I have some ideas of how to improve that situation, but I ran into issues with Couch throwing errors, so I may wait to optimize memory usage until after the Couch to SQL migration is complete. In any case caching rows did not cause memory usage to change too much.

millerdev commented on corehq/apps/fixtures/upload/workbook.py:
```diff
@@ -70,20 +72,27 @@ def iter_tables(self, domain):
-            yield FixtureDataItem(
+        sort_key = -1
+        for i, di in enumerate(data_items):
+            sort_key = max(sort_keys.get(di['UID'], i), sort_key + 1)
```
`sort_keys` is a dict of `{<row UUID>: <old sort key (int)>}`

> Is it correct to say it was previously sorted by the order of the rows in the sheet and now is sorted by the unique id of the row, which means it was user controlled and now is not?

No. It is still user-controlled, and now it is more correct than it was before because the order of the rows in the spreadsheet always determine the sort order where before that was only true when using `replace=True` (in which case all rows were replaced with new rows).

Previously, uploading with `replace=False` would cause new rows to be assigned sort keys that matched existing rows, which would cause old and new rows to be intermixed.

millerdev commented on corehq/apps/fixtures/upload/run_upload.py:
```diff
@@ -550,7 +549,7 @@ def _process_group_ownership(di, old_data_item, group_memoizer, transaction):
 
     for group_name in di.get('group', []):
-        group = group_memoizer.by_name(group_name)
+        group = group_memoizer.by_name(str(group_name))
```
This is now obsolete (it has been removed with the old uploader), but it handled a non-string group name value. This can happen when non-string values (such as numbers) are entered in the Excel sheet. A non-string group name would cause the old uploader to crash.

calellowitz commented on corehq/apps/fixtures/upload/run_upload.py:
```diff
@@ -45,10 +45,22 @@ def _run_upload(domain, workbook, replace=False, task=None):
-        mutation = get_mutation(old_rows, new_rows, row_key)
-        rows.update(mutation)
+        rows_by_id = {r._id: r for r in old_rows}
+        assert None not in rows_by_id, rows_by_id[None]
```
Yeah, sorry I understand how assertions are used. I meant that we often use formatted strings for the assertion message (you can see examples of that in sentry now), which makes it easier to understand what the assertion was supposed to mean in the unlikely event there is a bug and the assertion is triggered.

calellowitz replied:
Sure, I think messages in an assertion like https://github.com/dimagi/commcare-hq/blob/master/corehq/util/doc_processor/couch.py#L167 make it a bit easier to see exactly what is wrong, especially for someone other than the author of the assertion, since they describe the issue rather than just showing the `repr` of a `FixtureDataItem` class, which may not be immediately obvious to someone else who finds that error in sentry.

This is definitely not a blocker, but I do think it's a nice convention that shows up frequently in our code and makes it a lot easier if someone other than the original author stumbles upon an `AssertionError`

millerdev committed:
Update progress at most every 2s

millerdev committed:
Flush accumulated rows to db when count > 1000

Improves memory consumption for large workbooks. The retries are a work around for Couch errors seen while profiling: Internal Server Error timeout The request could not be processed in a reasonable amount of time.

millerdev committed:
Merge branch 'master' into dm/lookup-table-uploader
